### PR TITLE
added REST call for customer synchronization

### LINKED ISSUE #10: Add calls for client put and delete for automatic scripts.
Add REST HTTP call `sync_customer_tf`

*HTTP Methods*:

    - **GET**:
            - accepts request args as filters for table fields
    - **PUT**:
            - accepts *data* field; treated as JSON; fields:
                - `code` (str, mandatory)
                - `country` (str, mandatory)
                - `language` (str, optional)
                - `language_id` (int, optional)
                - `is_active` (bool, optional, default=`True`)
            - `code` field is a primary key; update all other fields from request if such record present; create new one if not; `language_id` should be a key to `ClientLanguage` record by following algorithm:
                - table_record.id == request.language_id; continue if not found;
                - table_record.code == request.language (case insensitive); continue if not found;
                - table_record.description == request.language (case insensitive); continue if not found;
                - table_record.code == default_code;
            - check corresponding record in `FtpUploadOptions` table; create default one if not (`client_id = client.id, should_encrypt = True, can_receive = True);
            - Response code:
                - HTTP 400 (bad request) in case of any error, text part is a text representation of an exception raised;
                - HTTP 201 (created) - if success, see text JSON description below;
        - **DELETE**
            - accepts *data* field, but `code` key is mandatory only; others should be ignored;
            - found `Client` record and save `is_active = False` flag; do nothing if no such record found;
            - response code:
                - HTTP 400 (bad request) in case of any error, text part is a text representation of an exception raised;
                - HTTP 200 (ok) - if success, see text JSON description below;

*Response JSON (for success)*:

      - Header required: `Content-type: application/json`;
      - See response code in HTTP methods specification above;
      - Filter by request arguments if supplied; return all active clients otherwise;
      - Sort all keys inside hashes; sorting have to be the same always;
      - Do not convert data provided by database, that is: do not change type, case, e.t.c.
      - list itself should be sorted by `code` field ascending (alphabetically)
```
        [
            {
                "id": <client.id, int>,
                "code" : <client.code, str>,
                "country": <client.country, str>,
                "is_active": <client.is_active, bool>,
                "language_id": <client.language_id>
            },
           .....
           .....
        ]
```

### PR DESCRIPTION
**Pull request by kriskutuzovaa**

### CODE REVIEW

**Comment by kriskutuzovaa** [comment]
@vifrrg

File: setup.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 from setuptools import setup
 
-__version = "1.0.4"
+__version = "1.0.5"
```
Comment (vifrrg) [inline setup.py]: Changes are enough to update `minor` component:
`1.1.0`

File: oc_client_provider/app/client_tf.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            _clients = Client.objects.filter(**search_params).order_by('code')
+        except Client.DoesNotExist as _e:
+            _clients = list()
+
+        _result = list()
+        for _record in _clients:
+            _record = _record.__dict__
+            _result.append(dict((__k, _record[__k]) for __k in sorted(_record.keys()) if not __k.startswith('_')))
+
+        logging.debug("Returning '%s'" % str(_result))
```
Comment (vifrrg) [inline oc_client_provider/app/client_tf.py:34]: Please be consistent.
You used `json.dumps` previously for the same reason.

File: oc_client_provider/app/client_tf.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        return _result
+
+    def put_client(self, **client):
+        """
+        Put new client from data
+        :param client: client data
+        :return dict: list of client data or None
```
Comment (vifrrg) [inline oc_client_provider/app/client_tf.py]: This routine returns nothing actually.
Please adjust specification.

File: oc_client_provider/app/client_tf.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        if _ftp_created:
+            _ftp_options.save()
+
+        logging.debug('Сlient with code [%s] created successfully' % _client.code)
+
+    def delete_client(self, **client):
+        """
+        Delete client given by client code
+        :param client: client data
+        :return dict: list of client data or None
```
Comment (vifrrg) [inline oc_client_provider/app/client_tf.py]: This routine returns nothing actually.
Please adjust specification.

File: oc_client_provider/app/routes.py [hunk 5]
```diff
@@ -180,10 +181,32 @@ def get_client_data (client_id):
...
     return response_json(200, {client_code: ClientCounterparty().client_counterparty(client_code)})
+
+@client_provider_bp.route('/sync_customer_tf', methods=['GET', 'PUT', 'DELETE'])
+def sync_customer_tf():
+    """
+    Synchronize client record and return filtered list as requested
+    """
+    # doing our business depending on request type
+    # returning normal JSON response
+    logging.info("/sync_customer_tf")
```
Comment (vifrrg) [inline oc_client_provider/app/routes.py]: Please add requestor's ip address to this message. You may find it in `request` object.

File: oc_client_provider/tests/test_client_provider.py [hunk 6]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
         self.assertEqual(len(_response.json), 1)
         _delivery = _response.json.pop()
         self.assertTrue(_delivery.get('creation_date_mr').startswith(_day_to_check))
+
+    def test_get_client_data_list_tf(self):
+        response = self.test_client.get("/sync_customer_tf")
+        self.assertEqual(response.status_code, 200)
+        _actual_dict = response.json
+
+        _expected_dict = list()
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: *IMHO*: bad idea to call a variable `dict` if it is `list` actually.

File: oc_client_provider/tests/test_client_provider.py [hunk 7]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
+        for _record in dl_models.Client.objects.filter(is_active=True).order_by('code'):
+            _record = _record.__dict__
+            _expected_dict.append(dict((__k, _record[__k]) for __k in sorted(_record.keys()) if not __k.startswith('_')))
+
+        self.assertEqual(_actual_dict, _expected_dict)
+        self.assertEqual(len(_actual_dict), 51)
+
+    def test_get_client_data_list_tf__with_parameters(self):
+        self.test_client.put("/sync_customer_tf",
+                             json={'code': 'TEST_CLIENT_999', 'country': 'Scotland', 'language_id': 2})
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: Please do not use real objects (including countries) in test data.

File: oc_client_provider/tests/test_client_provider.py [hunk 8]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
+
+        _expected_record = dl_models.Client.objects.get(code='TEST_CLIENT_999').__dict__
+        _expected_record = dict((__k, _expected_record[__k]) for __k in sorted(_expected_record.keys())
+                                if not __k.startswith('_'))
+        _expected_dict = [_expected_record]
+
+        response = self.test_client.get("/sync_customer_tf",
+                                        query_string={'id': _expected_record.get('id'), 'code': 'TEST_CLIENT_999',
+                                                      'country': 'Scotland', 'language_id': 2})
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: **Crime**: real country in test data

File: oc_client_provider/tests/test_client_provider.py [hunk 9]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
+        self.assertEqual(response.status_code, 200)
+        _actual_dict = response.json
+
+        self.assertEqual(_actual_dict, _expected_dict)
+        self.assertEqual(len(_actual_dict), 1)
+
+    def test_delete_client_data_tf(self):
+        response = self.test_client.delete("/sync_customer_tf", json={'code': 'TEST_CLIENT_1', 'lala': 'lala'})
+        self.assertEqual(response.status_code, 200)
+        _actual_dict = response.json
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: `dict` variable is `list` actually

File: oc_client_provider/tests/test_client_provider.py [hunk 10]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
+        _expected_dict = list()
+        for _record in dl_models.Client.objects.filter(is_active=True).order_by('code'):
+            _record = _record.__dict__
+            _expected_dict.append(dict((__k, _record[__k]) for __k in sorted(_record.keys()) if not __k.startswith('_')))
+
+        self.assertEqual(_actual_dict, _expected_dict)
+        self.assertEqual(len(_actual_dict), 50)
+
+    def test_put_client_data_tf(self):
+        response = self.test_client.put("/sync_customer_tf", json={'code': 'TEST_CLIENT_999', 'country': 'Scotland'})
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: real country in test data

File: oc_client_provider/tests/test_client_provider.py [hunk 11]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
+        response = self.test_client.put("/sync_customer_tf",
+                                        json={'code': 'TEST_CLIENT_999',
+                                              'country': 'Scotland',
+                                              'is_active': False,
+                                              'language': 'en'},
+                                        query_string={'is_active': False})
+        self.assertEqual(response.status_code, 201)
+        _actual_dict = response.json
+
+        _expected_dict = list()
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: `dict` is `list`...

File: oc_client_provider/tests/test_client_provider.py [hunk 12]
```diff
@@ -293,10 +295,85 @@ def test_get_deliveries_v2__date_range(self):
...
+            _record = _record.__dict__
+            _expected_dict.append(dict((__k, _record[__k]) for __k in sorted(_record.keys()) if not __k.startswith('_')))
+
+        self.assertEqual(_actual_dict, _expected_dict)
+        self.assertEqual(len(_actual_dict), 52)
+
+    def test_put_client_data_tf__with_parameters(self):
+        response = self.test_client.put("/sync_customer_tf",
+                                        json={'code': 'TEST_CLIENT_999',
+                                              'country': 'Scotland',
```
Comment (vifrrg) [inline oc_client_provider/tests/test_client_provider.py]: real country in test data

**Review by vifrrg — CHANGES_REQUESTED** [review]
Please update version and get rid of real objects and names in test data.

File: oc_client_provider/app/client_tf.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                logging.exception(_e)
+                pass
+
+            try:
+                return list(ClientLanguage.objects.filter(description__iexact=language)).pop()
+            except IndexError as _e:
+                logging.exception(_e)
+                pass
+
+        return ClientLanguage.objects.get(code='rus')
```
Comment (vifrrg) [inline oc_client_provider/app/client_tf.py]: Please parametrize default language, shoul be English if not set.

**Review by vifrrg — APPROVED** [review]

**Pull request merged**